driver-core/src/main/com/mongodb/connection/DefaultConnectionPool.java

```diff
@@ -83,10 +83,10 @@ public InternalConnection get() {
     @Override
     public InternalConnection get(final long timeout, final TimeUnit timeUnit) {
         try {
+            connectionPoolListener.waitQueueEntered(new ConnectionPoolWaitQueueEvent(serverId, currentThread().getId()));
```

Comment:
I think this is incorrect.  The thread can't enter the wait queue if the wait queue is full, so the event should not be generated.

How about instead we pull the throw of the wait queue full exception up out of the try block? That way the finally block will not be executed when the exception is thrown.